springboard/views.py

```diff
@@ -80,3 +87,60 @@ def api_notify(self):
     def notfound(self):
         self.request.response.status = 404
         return {}
+
+    @view_config(
+        route_name='locale_change',
+        renderer='springboard:templates/locale_change.jinja2')
+    def locale_change(self):
+        return {
+            'languages': self.get_featured_languages() +
+            sorted(list(set(self.get_available_languages()) -
+                        set(self.get_featured_languages())),
+                   key=lambda tup: tup[1].lower())
+        }
+
+    @view_config(route_name='locale')
+    @view_config(route_name='locale_matched')
+    def set_locale_cookie(self):
+        response = Response()
+        language = self.request.matchdict.get('language') or \
+            self.request.GET.get('language')
+
+        if language:
+            response.set_cookie('_LOCALE_', value=language, max_age=31536000)
+
+        return HTTPFound(location='/', headers=response.headers)
+
+    def get_available_languages(self):
+        available_languages = sorted(literal_eval(
+            (self.settings.get('available_languages', '[]'))),
+            key=lambda tup: tup[1].lower())
+        return [
+            (code, self.get_display_name(code))
+            for code, name in available_languages]
+
+    def get_featured_languages(self):
+        featured_languages = sorted(literal_eval(
+            (self.settings.get('featured_languages', '[]'))),
+            key=lambda tup: tup[1].lower())
+        return [
+            (code, self.get_display_name(code))
+            for code, name in featured_languages]
+
+    def get_display_name(self, locale):
```

Comment:
I think we should only use the locale codes in view logic and leave it to the template to convert it to a display name